File: Linked List/0025.Reverse Nodes in k-Group/Solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if k == 1:
            return head
        tail, tmp = head, head
        vals = []
        cnt = 0
        while tail:
            vals.append(tail.val)
            cnt += 1
            if cnt == k:
                while vals:
                    tmp.val = vals.pop()
                    tmp = tmp.next
                cnt = 0
                tmp = tail.next
            tail = tail.next
        return head

File: Linked List/0025.Reverse Nodes in k-Group/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if (k == 1) {
            return head;
        }
        ListNode tail = head, tmp = head;
        Deque<Integer> vals = new ArrayDeque<>();
        int cnt = 0;
        while (tail != null) {
            vals.addLast(tail.val);
            if (++cnt == k) {
                while (!vals.isEmpty()) {
                    tmp.val = vals.pollLast();
                    tmp = tmp.next;
                }
                cnt = 0;
                tmp = tail.next;
            }
            tail = tail.next;
        }
        return head;
    }
}